2019-02-07  Ethan W Kwan  <ethank@lnxsrv06.seas.ucla.edu>

        * timespec.h: Uses direct comparison instead of integer subtraction.
        If seconds are equivalent, return the difference of nanoseconds.
        This way, signed overflow is prevented.


--- coreutils-with-bug/lib/timespec.h	2005-09-21 23:05:39.000000000 -0700
+++ lib/timespec.h	2019-02-07 22:25:03.189158000 -0800
@@ -45,8 +45,17 @@
 static inline int
 timespec_cmp (struct timespec a, struct timespec b)
 {
-  int diff = a.tv_sec - b.tv_sec;
-  return diff ? diff : a.tv_nsec - b.tv_nsec;
+  if(a.tv_sec < b.tv_sec){
+    return -1;
+  }
+
+  else if(a.tv_sec > b.tv_sec){
+    return 1;
+  }
+
+  else {
+    return a.tv_nsec - b.tv_nsec;
+  }
 }
 
 # if ! HAVE_DECL_NANOSLEEP
